multik-cuda/src/main/kotlin/org/jetbrains/kotlinx/multik/cuda/GpuArray.kt

```diff
@@ -0,0 +1,125 @@
+import jcuda.runtime.cudaError.cudaErrorMemoryAllocation
+import jcuda.runtime.cudaMemcpyKind
+import mu.KotlinLogging
+import org.jetbrains.kotlinx.multik.ndarray.data.Dimension
+import org.jetbrains.kotlinx.multik.ndarray.data.MultiArray
+import java.lang.ref.Cleaner
+import java.util.*
+import java.util.concurrent.LinkedBlockingQueue
+
+
+private val logger = KotlinLogging.logger {}
+
+internal class GpuArray private constructor(
+    private val rawData : Any,
+    val hostDataPtr : Pointer,
+    val deviceDataPtr : Pointer,
+    val byteSize : Long
+) {
+    var isLoaded = true
+        private set
+
+    fun copyFromGpu() {
+        checkResult(JCuda.cudaMemcpy(hostDataPtr, deviceDataPtr, byteSize, cudaMemcpyKind.cudaMemcpyDeviceToHost))
+    }
+
+    private fun free() {
+        if (!isLoaded)
+            throw IllegalStateException("Trying to free memory that it is already freed")
+
+        logger.debug { "Freeing GPU memory. Data: $rawData, size: ${byteSizeToString(byteSize)}" }
+        checkResult(JCuda.cudaFree(deviceDataPtr))
+
+        isLoaded = false
+    }
+
+    companion object {
+        private const val CACHE_INITIAL_CAPACITY = 16
+        private const val CACHE_LOAD_FACTOR = 0.75f
+
+        private val cache = Collections.synchronizedMap(
+            LinkedHashMap<Any, GpuArray>(CACHE_INITIAL_CAPACITY, CACHE_LOAD_FACTOR, true)
+        )
+
+        private val deleteQueue = LinkedBlockingQueue<Any>()
+        private val cleaner = Cleaner.create()
+
+        fun assertAllLoaded(vararg arrays: GpuArray) {
+            if (!arrays.all { it.isLoaded })
+                throw OutOfMemoryError("Not all arrays have are loaded in the GPU memory")
+        }
+
+        fun fullCleanup() {
+            cache.forEach { it.value.free() }
+            cache.clear()
+
+            deleteQueue.clear()
+        }
+
+        fun <T : Number, D : Dimension> getOrAlloc(array: MultiArray<T, D>, setMemory: Boolean = true): GpuArray {
+            cleanup()
+            return cache.getOrPut(array.data.data) { allocMemory(array, setMemory) }
+        }
+
+        private fun getGpuMemInfo(): String {
+            val free = LongArray(1)
+            val total = LongArray(1)
+
+            checkResult(JCuda.cudaMemGetInfo(free, total))
+            return "free: ${byteSizeToString(free[0])}, total: ${byteSizeToString(total[0])} MB"
+        }
+
+        private fun <T : Number, D : Dimension> allocMemory(array: MultiArray<T, D>, setMemory: Boolean): GpuArray {
+            val elemSize = array.dtype.itemSize
+            val rawData = array.data.data
+
+            val deviceDataPtr = Pointer()
+            val hostDataPtr = array.dtype.getDataPointer(array.data)
+            val byteSize = elemSize.toLong() * array.size
+
+            logger.debug { "Allocating array on GPU. Data: ${rawData}, size: ${byteSizeToString(byteSize)}" }
+
+            while (true) {
+                val result = JCuda.cudaMalloc(deviceDataPtr, byteSize)
```

Comment:
Isn't it possible to get information about memory and its lack before?

Reply:
Changed allocation logic to check for free memory before trying to malloc